File: RESTFul/AAB/PHP/app1/lib/AAB/AABService.php
<?php
namespace Att\Api\AAB;

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * AAB Library
 *
 * PHP version 5.4+
 *
 * LICENSE: Licensed by AT&T under the 'Software Development Kit Tools
 * Agreement.' 2014.
 * TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTIONS:
 * http://developer.att.com/sdk_agreement/
 *
 * Copyright 2014 AT&T Intellectual Property. All rights reserved.
 * For more information contact developer.support@att.com
 *
 * @category  API
 * @package   AAB
 * @author    pk9069
 * @copyright 2014 AT&T Intellectual Property
 * @license   http://developer.att.com/sdk_agreement AT&amp;T License
 * @link      http://developer.att.com
 */

require_once __DIR__ . '../../Srvc/APIService.php';
require_once __DIR__ . '/../Restful/HttpPatch.php';
require_once __DIR__ . '/Contact.php';
require_once __DIR__ . '/ContactCommon.php';
require_once __DIR__ . '/ContactsResultSet.php';
require_once __DIR__ . '/GroupsResultSet.php';

use InvalidArgumentException;
use Att\Api\OAuth\OAuthToken;
use Att\Api\Restful\HttpGet;
use Att\Api\Restful\HttpPatch;
use Att\Api\Restful\HttpPost;
use Att\Api\Restful\RestfulRequest;
use Att\Api\Srvc\APIService;
use Att\Api\Srvc\Service;
use Att\Api\Srvc\ServiceException;

/**
 * Used to interact with version 1 of the Address Book API Gateway.
 *
 * @category API
 * @package  AAB
 * @author   pk9069
 * @license  http://developer.att.com/sdk_agreement AT&amp;T License
 * @version  Release: @package_version@
 * @link     https://developer.att.com/docs/apis/rest/1/Advertising
 */
class AABService extends APIService
{

    /**
     * Creates an AABService object that can be used to interact with
     * the Address Book API Gateway.
     *
     * @param string     $fqdn  fully qualified domain name to which requests
     *                          will be sent
     * @param OAuthToken $token OAuth token used for authorization
     */
    public function __construct($fqdn, OAuthToken $token)
    {
        parent::__construct($fqdn, $token);
    }

    /**
     * Sends a request to the API gateway for creating a contact.
     *
     * @param ContactCommon $contact contact information to create
     *
     * @return string location of created resource
     * @throws ServiceException if request was not successful
     */
    public function createContact(ContactCommon $contact)
    {
        $endpoint = $this->getFqdn() . '/addressBook/v1/contacts';

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Content-Type', 'application/json');

        $httpPost = new HttpPost();

        $arrBody = array('contact' => $contact->toArray());
        $httpPost->setBody(json_encode($arrBody));

        $result = $req->sendHttpPost($httpPost);

        $code = $result->getResponseCode();

        if ($code != 201 && $code != 204) {
            throw new ServiceException($result->getResponseBody(), $code);
        }

        return $result->getHeader('location');
    }

    /**
     * Sends a request to the API gateway for getting a contact.
     *
     * @param string      $contactId contact id used to get contact
     * @param string|null $xFields   x-fields to send or null if none
     *
     * @return Contact|QuickContact returned contact
     * @throws ServiceException if request was not successful
     */
    public function getContact($contactId, $xFields=null)
    {
        $subUrl = '/addressBook/v1/contacts/' . $contactId;
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        if ($xFields != null) {
            $req->setHeader('x-fields', $xFields);
        }

        $result = $req->sendHttpGet();

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);

        if (isset($arr['quickContact'])) {
            return ContactCommon::fromArray($arr);
        } else {
            return Contact::fromArray($arr);
        }
    }

    /**
     * Sends a request to the API gateway for getting contacts.
     *
     * @param PaginationParameters $pParams paginiations parameters
     * @param string|null          $search  search text
     *
     * @return ContactResultSet contact result set
     * @throws ServiceException if request was not successful
     */
    public function getContacts($xFields=null,
        PaginationParameters $pParams=null, $search=null
    ) {

        $endpoint = $this->getFqdn() . '/addressBook/v1/contacts';

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        if ($xFields != null) {
            $req->setHeader('x-fields', $xFields);
        }

        $httpGet = new HttpGet();

        if ($pParams != null) {
            $httpGet->setParams($pParams->toArray());
        }
        if ($search != null) {
            $httpGet->setParam('search', $search);
        }

        $result = $req->sendHttpGet($httpGet);

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);

        return ContactsResultSet::fromArray($arr);
    }

    /**
     * Sends a request to the API gateway for getting contact groups.
     *
     * @param string               $contactId contact id to get groups for
     * @param PaginationParameters $params    paginiations parameters
     *
     * @return GroupsResultSet returns contact groups
     * @throws ServiceException if request was not successful
     */
    public function getContactGroups(
        $contactId, PaginationParameters $params=null
    ) {

        $subUrl = "/addressBook/v1/contacts/$contactId/groups";
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $httpGet = new HttpGet();
        if ($params != null) {
            $httpGet->setParams($params->toArray());
        }

        $result = $req->sendHttpGet($httpGet);

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);

        return GroupsResultSet::fromArray($arr);
    }

    /**
     * Sends a request to the API gateway for updating a contact.
     *
     * @param Contact $contact contact to update contact id used to get contact
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function updateContact(Contact $contact)
    {
        if ($contact->getContactId() == null) {
            throw new InvalidArgumentException('Contact id must not be null.');
        }

        $contactId = urlencode($contact->getContactId());
        $subUrl = '/addressBook/v1/contacts/' . $contactId;
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $body = json_encode(array('contact' => $contact->toArray()));
        $httpPatch = new HttpPatch($body);
        $result = $req->sendHttpPatch($httpPatch);

        $code = $result->getResponseCode();

        if ($code != 201 && $code != 204) {
            throw new ServiceException($code, $result->getResponseBody());
        }
    }

    /**
     * Sends a request to the API gateway for deleting a contact.
     *
     * @param string $contactId contact id to delete
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function deleteContact($contactId)
    {
        $fqdn = $this->getFqdn();
        $endpoint = $fqdn . '/addressBook/v1/contacts/' . $contactId;

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json');

        $result = $req->sendHttpDelete();

        $code = $result->getResponseCode();

        if ($code != 201 && $code != 204) {
            throw new ServiceException($code, $result->getResponseBody());
        }
    }

    /**
     * Sends a request to the API gateway for creating a group.
     *
     * @param Group $group group information to create
     *
     * @return string location of created resource
     * @throws ServiceException if request was not successful
     */
    public function createGroup(Group $group)
    {
        $endpoint = $this->getFqdn() . '/addressBook/v1/groups';

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $httpPost = new HttpPost();
        $httpPost->setBody(json_encode($group->toArray()));

        $result = $req->sendHttpPost($httpPost);

        $code = $result->getResponseCode();
        if ($code != 201 && $code != 204) {
            throw new ServiceException($code, $result->getResponseBody());
        }
        return $result->getHeader('location');
    }

    /**
     * Sends a request to the API gateway for getting groups.
     *
     * @param PaginationParameters $pParams   paginiations parameters
     * @param string|null          $groupName group name used in search, if any
     *
     * @return ResultSet returned groups
     * @throws ServiceException if request was not successful
     */
    public function getGroups(PaginationParameters $pParams=null,
        $groupName=null
    ) {
        $endpoint = $this->getFqdn() . '/addressBook/v1/groups';

        $httpGet = new HttpGet();

        if ($pParams != null) {
            $httpGet->setParams($pParams->toArray());
        }
        if ($groupName != null) {
            $httpGet->setParam('groupName', $groupName);
        }

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $result = $req->sendHttpGet($httpGet);

        if ($result->getResponseCode() == 204) {
            return null;
        }

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);

        return GroupsResultSet::fromArray($arr);
    }

    /**
     * Sends a request to the API gateway for deleting a group.
     *
     * @param string $groupId group id to delete
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function deleteGroup($groupId)
    {
        $groupId = urlencode($groupId);
        $endpoint = $this->getFqdn() . "/addressBook/v1/groups/$groupId";

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $result = $req->sendHttpDelete();
        $code = $result->getResponseCode();
        if ($code != 204) {
            throw new ServiceException($result->getResponseBody(), $code);
        }
    }

    /**
     * Sends a request to the API gateway for updating a group.
     *
     * @param Group $group group to update
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function updateGroup(Group $group)
    {
        if ($group->getGroupId() == null) {
            throw new InvalidArgumentException('Group id must not be null.');
        }

        $subUrl = '/addressBook/v1/groups/' . urlencode($group->getGroupId());
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RESTFulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $body = json_encode($group->toArray());
        $result = $req->sendHttpPatch(new HttpPatch($body));

        $code = $result->getResponseCode();
        if ($code != 204) {
            throw new ServiceException($result->getResponseBody(), $code);
        }
    }

    /**
     * Sends a request to the API gateway adding contacts to a group.
     *
     * @param string $groupId    group id to add contacts to
     * @param array  $contactIds contact ids to add to group
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function addContactsToGroup($groupId, $contactIds)
    {
        $groupId = urlencode($groupId);
        $contactIds = urlencode(implode(',', $contactIds));

        $subUrl =
            "/addressBook/v1/groups/$groupId/contacts?contactIds=$contactIds";
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RESTFulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $result = $req->sendHttpPost();

        $code = $result->getResponseCode();
        if ($code != 204) {
            throw new ServiceException($result->getResponseBody(), $code);
        }
    }

    /**
     * Sends a request to the API gateway removing contacts from a group.
     *
     * @param string $groupId    group id to remove contacts from
     * @param array  $contactIds contact ids remove from group
     *
     * @return void
     * @throws ServiceException if request was not successful
     */
    public function removeContactsFromGroup($groupId, $contactIds)
    {
        $groupId = urlencode($groupId);
        $contactIds = urlencode(implode(',', $contactIds));

        $subUrl =
            "/addressBook/v1/groups/$groupId/contacts?contactIds=$contactIds";
        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RESTFulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $result = $req->sendHttpDelete();

        $code = $result->getResponseCode();
        if ($code != 204) {
            throw new ServiceException($result->getResponseBody(), $code);
        }
    }

    /**
     * Sends a request to the API gateway for getting contacts owned by a group.
     *
     * @param string               $groupId group id
     * @param PaginationParameters $params  paginiation params
     *
     * @return array contact ids
     * @throws ServiceException if request was not successful
     */
    public function getGroupContacts($groupId, PaginationParameters $params=null)
    {
        $groupId = urlencode($groupId);
        $subUrl = "/addressBook/v1/groups/$groupId/contacts";

        $endpoint = $this->getFqdn() . $subUrl;

        $req = new RESTFulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json');

        $httpGet = new HttpGet();
        if ($params != null) {
            $httpGet->setParams($params->toArray());
        }

        $result = $req->sendHttpGet($httpGet);

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);
        $contactName = $arr['contactIds'];
        $contactArr = $contactName['id'];

        return $contactArr;
    }

    /**
     * Sends a request to the API gateway for getting subscriber's personal
     * contact card.
     *
     * @return Contact contact information for subscriber
     * @throws ServiceException if request was not successful
     */
    public function updateMyInfo(ContactCommon $contact)
    {
        $endpoint = $this->getFqdn() . '/addressBook/v1/myInfo';

        $req = new RestfulRequest($endpoint);

        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');

        $body = json_encode(array('myInfo' => $contact->toArray()));
        $httpPatch = new HttpPatch($body);
        $result = $req->sendHttpPatch($httpPatch);

        $code = $result->getResponseCode();

        if ($code != 201 && $code != 204) {
            throw new ServiceException($code, $result->getResponseBody());
        }
    }

    /**
     * Sends a request to the API gateway for getting subscriber's personal
     * contact card.
     *
     * @return Contact contact information for subscriber
     * @throws ServiceException if request was not successful
     */
    public function getMyInfo()
    {
        $endpoint = $this->getFqdn() . '/addressBook/v1/myInfo';

        $req = new RESTFulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json');

        $result = $req->sendHttpGet();

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);

        return Contact::fromArray($arr['myInfo']);
    }

}

?>

File: RESTFul/AAB/Java/app1/src/main/java/com/att/api/aab/service/AABService.java
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.att.api.aab.service;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.att.api.oauth.OAuthToken;
import com.att.api.rest.APIResponse;
import com.att.api.rest.RESTClient;
import com.att.api.rest.RESTException;
import com.att.api.service.APIService;

/**
 * Used to interact with version 1 of the Address Book API.
 *
 * <p>
 * This class is thread safe.
 * </p>
 *
 */
public class AABService extends APIService {
    
    private void addPageParams(RESTClient client, PageParams params) {
        if (params == null) return;

        final String[] keys = { "order", "orderBy", "limit", "offset" };
        final String[] values = { 
            params.getOrder(), params.getOrderBy(), params.getLimit(),
            params.getOffset() 
        };

        for (int i = 0; i < keys.length; ++i) {
            if (values[i] == null) continue;
            client.addParameter(keys[i], values[i]);
        }
    }

    /**
     * Creates a AABService object.
     *
     * @param fqdn fully qualified domain name to use for sending requests
     * @param token OAuth token to use for authorization
     */
    public AABService(final String fqdn, final OAuthToken token) {
        super(fqdn, token);
    }

    public String createContact(Contact contact) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/contacts";

        JSONObject payload = new JSONObject();
        payload.put("contact", contact.toJson());
        APIResponse response = new RESTClient(endpoint)
            .addHeader("Content-Type", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPost(payload.toString());

        if (response.getStatusCode() != 201) {
            throw new RESTException(response.getResponseBody());
        }

        return response.getHeader("location");
    }

    public ContactWrapper getContact(String contactId, String xFields)
            throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/contacts/" + contactId;

        RESTClient client = new RESTClient(endpoint)
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken());

        if (xFields != null) {
            client.addHeader("x-fields", xFields);
        }

        APIResponse response = client.httpGet();
        if (response.getStatusCode() != 200) {
            throw new RESTException(response.getResponseBody());
        }

        final String body = response.getResponseBody();
        JSONObject jobj = new JSONObject(body);
        if (jobj.has("contact")) {
            Contact contact = Contact.valueOf(jobj.getJSONObject("contact"));
            return new ContactWrapper(contact);
        }

        JSONObject jQc = jobj.getJSONObject("quickContact");
        return new ContactWrapper(QuickContact.valueOf(jQc));
    }


    public ContactResultSet getContacts() throws RESTException {
        return getContacts(null, null, null);
    }

    public ContactResultSet getContacts(String sParams) throws RESTException {
        return getContacts(null, null, sParams);
    }

    public ContactResultSet getContacts(String xFields, String sParams) 
            throws RESTException {
        return getContacts(xFields, null, sParams);
    }

    public ContactResultSet getContacts(PageParams pParams, String sParams)
            throws RESTException {
        return getContacts(null, pParams, sParams);
    }

    public ContactResultSet getContacts(String xFields, PageParams pParams,
            String sParams) throws RESTException {

        String endpoint = getFQDN() + "/addressBook/v1/contacts";

        RESTClient client = new RESTClient(endpoint)
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken());

        if (xFields != null) {
            client.addHeader("x-fields", xFields);
        }
        if (pParams != null) {
            this.addPageParams(client, pParams);
        }

        if (sParams != null) {
            client.addParameter("search", sParams);
        }

        APIResponse response = client.httpGet();
        if (response.getStatusCode() != 200) {
            throw new RESTException(response.getResponseBody());
        }
        final String body = response.getResponseBody();
        try {
            JSONObject jrs = new JSONObject(body).getJSONObject("resultSet");
            return ContactResultSet.valueOf(jrs);
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public GroupResultSet getContactGroups(String contactId)
            throws RESTException {
        return getContactGroups(contactId, null);
    }

    public GroupResultSet getContactGroups(String contactId, PageParams params)
            throws RESTException {
        String endpoint = getFQDN() 
            + "/addressBook/v1/contacts/" + contactId + "/groups";

        RESTClient client = new RESTClient(endpoint)
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken());

        this.addPageParams(client, params);
        APIResponse response = client.httpGet();
        if (response.getStatusCode() != 200) {
            throw new RESTException(response.getResponseBody());
        }
        final String body = response.getResponseBody();
        try {
            JSONObject jrs = new JSONObject(body).getJSONObject("resultSet");
            return GroupResultSet.valueOf(jrs);
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public void updateContact(Contact contact, String contactId)
            throws RESTException {

        String endpoint = getFQDN() + "/addressBook/v1/contacts/" + contactId;

        JSONObject payload = new JSONObject();
        payload.put("contact", contact.toJson());
        APIResponse response = new RESTClient(endpoint)
            .addHeader("Content-Type", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPatch(payload.toString());

        int statusCode = response.getStatusCode();
        if (statusCode != 200 && statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public void deleteContact(String contactId) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/contacts/" + contactId;

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .httpDelete();

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public String createGroup(Group group) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/groups";

        JSONObject payload = new JSONObject();
        payload.put("group", group.toJson());
        APIResponse response = new RESTClient(endpoint)
            .addHeader("Content-Type", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPost(payload.toString());

        int statusCode = response.getStatusCode();
        if (statusCode != 201) {
            throw new RESTException(response.getResponseBody());
        }

        return response.getHeader("location");
    }

    public GroupResultSet getGroups(String groupName) 
            throws RESTException {
        return getGroups(null, groupName);
    }

    public GroupResultSet getGroups(PageParams params, String groupName)
            throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/groups";

        RESTClient client = new RESTClient(endpoint)
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken());

        this.addPageParams(client, params);
        if (groupName != null) client.addParameter("groupName", groupName);
        APIResponse response = client.httpGet();

        int statusCode = response.getStatusCode();
        if (statusCode != 200) {
            throw new RESTException(response.getResponseBody());
        }

        final String body = response.getResponseBody();
        try {
            JSONObject jrs = new JSONObject(body).getJSONObject("resultSet");
            return GroupResultSet.valueOf(jrs);
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public void deleteGroup(String groupId) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/groups/" + groupId;

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .httpDelete();

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public void updateGroup(Group group, String groupId) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/groups/" + groupId;
        
        JSONObject payload = new JSONObject();
        payload.put("group", group.toJson());
        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .addHeader("Content-Type", "application/json")
            .httpPatch(payload.toString());

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public void addContactsToGroup(String groupId, String contactIds)
            throws RESTException {

        // TODO: encode contactIds
        String endpoint = getFQDN() + "/addressBook/v1/groups/" + groupId 
            + "/contacts?contactIds=" + contactIds;

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .httpPost();

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public void removeContactsFromGroup(String groupId, String contactIds)
            throws RESTException {

        // TODO: encode contactIds
        String endpoint = getFQDN() + "/addressBook/v1/groups/" + groupId 
            + "/contacts?contactIds=" + contactIds;

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .httpDelete();

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

    public String[] getGroupContacts(String groupId) throws RESTException {
        return getGroupContacts(groupId, null);
    }

    public String[] getGroupContacts(String groupId, PageParams params)
            throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/groups/" + groupId 
            + "/contacts";

        RESTClient client = new RESTClient(endpoint)
            .addHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken());

        this.addPageParams(client, params);
        APIResponse response = client.httpGet();

        int statusCode = response.getStatusCode();
        if (statusCode != 200) {
            throw new RESTException(response.getResponseBody());
        }

        final String body = response.getResponseBody();
        try {
            JSONObject jIds = new JSONObject(body).getJSONObject("contactIds");
            JSONArray idsArr = jIds.getJSONArray("id");
            String[] ids = new String[idsArr.length()];
            for (int i = 0; i < ids.length; ++i) {
                ids[i] = idsArr.getString(i);
            }
            return ids;
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public Contact getMyInfo() throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/myInfo";

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .addHeader("Accept", "application/json")
            .httpGet();

        int statusCode = response.getStatusCode();
        if (statusCode != 200) {
            throw new RESTException(response.getResponseBody());
        }

        final String body = response.getResponseBody();
        try {
            JSONObject jMyInfo = new JSONObject(body);
            return Contact.valueOf(jMyInfo.getJSONObject("myInfo"));
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public void updateMyInfo(Contact myInfo) throws RESTException {
        String endpoint = getFQDN() + "/addressBook/v1/myInfo";

        JSONObject payload = new JSONObject();
        payload.put("myInfo", myInfo.toJson());

        APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .addHeader("Content-Type", "application/json")
            .httpPatch(payload.toString());

        int statusCode = response.getStatusCode();
        if (statusCode != 204) {
            throw new RESTException(response.getResponseBody());
        }
    }

}
